dolphin/alert_manager/trap_receiver.py

```diff
@@ -43,6 +45,102 @@ def __init__(self, trap_receiver_address, trap_receiver_port,
+            auth_key = new_config.get("auth_key", None)
+            auth_protocol = new_config.get("auth_protocol", None)
+            privacy_key = new_config.get("privacy_key", None)
+            privacy_protocol = new_config.get("privacy_protocol", None)
+            if auth_key is not None:
+                auth_key = cryptor.decode(auth_key)
+            if privacy_key is not None:
+                privacy_key = cryptor.decode(privacy_key)
+            config.addV3User(
+                self.snmp_engine,
+                userName=username,
+                authKey=auth_key,
+                privKey=privacy_key,
+                authProtocol=self._get_usm_auth_protocol(ctxt,
+                                                         auth_protocol),
+                privProtocol=self._get_usm_priv_protocol(ctxt,
+                                                         privacy_protocol),
+                securityEngineId=v2c.OctetString(hexValue=engine_id))
+
+    def _delete_snmp_config(self, ctxt, snmp_config):
+        LOG.info("Delete snmp config:%s" % snmp_config)
+        version_int = self._get_snmp_version_int(ctxt, snmp_config["version"])
+        if version_int == constants.SNMP_V3_INT:
+            username = snmp_config['username']
+            engine_id = snmp_config['engine_id']
+            config.delV3User(self.snmp_engine, userName=username,
+                             securityEngineId=v2c.OctetString(
+                                 hexValue=engine_id))
+        else:
+            storage_id = snmp_config['storage_id']
+            community_index = self._get_community_index(storage_id)
+            config.delV1System(self.snmp_engine, community_index)
+
+    def _get_community_index(self, storage_id):
+        return storage_id.replace('-', '')
+
+    def _get_snmp_version_int(self, ctxt, version):
+        _version = version.lower()
+        version_int = constants.VALID_SNMP_VERSIONS.get(_version)
+        if version_int is None:
+            msg = "Invalid snmp version %s." % version
+            raise exception.InvalidSNMPConfig(detail=msg)
+
+        return version_int
+
+    def _get_usm_auth_protocol(self, ctxt, auth_protocol):
+        usm_auth_protocol = config.usmNoAuthProtocol
+        if auth_protocol is not None:
+            auth_protocol = auth_protocol.lower()
+            if auth_protocol == constants.SNMP_AUTH_PROTOCOL_SHA:
+                usm_auth_protocol = config.usmHMACSHAAuthProtocol
+            elif auth_protocol == constants.SNMP_AUTH_PROTOCOL_MD5:
+                usm_auth_protocol = config.usmHMACMD5AuthProtocol
+            else:
+                msg = "Invalid auth_protocol %s." % auth_protocol
+                raise exception.InvalidSNMPConfig(detail=msg)
+
+        return usm_auth_protocol
+
+    def _get_usm_priv_protocol(self, ctxt, privacy_protocol):
+        usm_priv_protocol = config.usmNoPrivProtocol
+        if privacy_protocol is not None:
```

Comment:
Can we keep a map and check the keys and value..This will avoid multiple if-else and also if new attribute adds then just update map. No change in other code
ex:
protocol_map = {constants.SNMP_PRIVACY_PROTOCOL_AES: config.usmAesCfb128Protocol, constants.SNMP_PRIVACY_PROTOCOL_DES: config.usmDESPrivProtocol}
if privacy_protocol is not None and privacy_protocol.lower() in protocol_map.keys():
      usm_priv_protocol = protocol_map.get(privacy_protocol.lower())
else:
     raise exception...

Reply:
Done